Make it possible to override permissions in django settings

eodolphi opened the pull request.

Commit:
Make it possible to override permissions in django settings

sergei-maertens commented:
Can you make sure all tests pass?

Thanks for the README update by the way, I've been neglecting that too much.

sergei-maertens commented on exportdb/permissions.py:
```diff
@@ -1,10 +1,11 @@
-def is_superuser(user):
-    return user.is_superuser
+def permission(*args, **kwargs):
+    predicate = getattr(settings, 'EXPORTDB_PERMISSION', rules.is_superuser)
```
I think we should rather do something like:

```
# in exportdb/conf.py
class ExportDBConf(AppConf):
    EXPORT_ROOT = os.path.join(settings.MEDIA_ROOT, EXPORT_SUBDIR)
    EXPORT_MEDIA_URL = posixpath.join(settings.MEDIA_URL, EXPORT_SUBDIR)
    EXPORT_CONF = {}
    PERMISSION = 'rules.is_superuser'  # add the default

    # form used in admin view to confirm export
    CONFIRM_FORM = 'django.forms.Form'
```

with

```
# in exportdb/permissions.py
...
from django.conf import settings

import rules

from .compat import import_string


@rules.predicate
def permission(*args, **kwargs):
    predicate = import_string(settings.EXPORTDB_PERMISSION)
    return predicate(*args, **kwargs)


rules.add_rule('exportdb.can_export', permission)
```

This should cover the desired flexibility for now and I'll have look at cleaner permission handling later. Rules is a very powerful library where you can logically combine predicates, we should definitely leverage that.

On a sidenote, please run pylint and PEP8 on the code :)

eodolphi replied:
I started out with something like this (without the appconf stuff, totally missed that). But we need `rules.is_group_member('something')` as a rule. Which will not work with `import_string`, that's why I choose this option.

sergei-maertens replied:
I'm assuming you're import `rules` in the settings file then?

sergei-maertens commented on tests/app/tests/test_permissions.py:
```diff
@@ -0,0 +1,41 @@
+
+        self.assertFalse(
+            rules.test_rule('exportdb.can_export', self.normal_user)
+        )
```
Maybe add a test here with a staff user too, just to be sure

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Fix styling. Add default for EXPORTDB_PERMISSION to appconf, add test

Commit:
Fix tests, depend on newer version of mock

eodolphi commented:
Tests pass now. I was using an older version of mock that was in my virtualenv by accident. So I updated that dependency to.

I added a default to the appconf, but did not change the setting to a string. See comment above. If that is an absolute requirement, I can change it, and work around it on our side.

sergei-maertens commented:
Let's just go with this version for now, we'll see if we can come up with a clever permissions systems later. I'll upload a minor update to PyPi in a minute.
